File: api/db/models/evaluation.py
from __future__ import annotations

from peewee import BigIntegerField, CharField, FloatField, IntegerField, TextField, ForeignKeyField

from api.db.base import DataBaseModel
from api.db.fields import JSONField
from api.db.models.dialog import Dialog


class EvaluationDataset(DataBaseModel):
    id = CharField(max_length=32, primary_key=True)
    tenant_id = CharField(max_length=32, null=False, index=True, help_text="tenant ID")
    name = CharField(max_length=255, null=False, index=True, help_text="dataset name")
    description = TextField(null=True, help_text="dataset description")
    kb_ids = JSONField(null=False, help_text="knowledge base IDs to evaluate against")
    created_by = CharField(max_length=32, null=False, index=True, help_text="creator user ID")
    create_time = BigIntegerField(null=False, index=True, help_text="creation timestamp")
    update_time = BigIntegerField(null=False, help_text="last update timestamp")
    status = IntegerField(null=False, default=1, help_text="1=valid, 0=invalid")

    class Meta:
        db_table = "evaluation_datasets"


class EvaluationCase(DataBaseModel):
    id = CharField(max_length=32, primary_key=True)
    dataset_id = ForeignKeyField(
        EvaluationDataset,
        column_name="dataset_id",
        null=False,
        index=True,
        help_text="FK to evaluation_datasets",
        on_delete="CASCADE",
    )
    question = TextField(null=False, help_text="test question")
    reference_answer = TextField(null=True, help_text="optional ground truth answer")
    relevant_doc_ids = JSONField(null=True, help_text="expected relevant document IDs")
    relevant_chunk_ids = JSONField(null=True, help_text="expected relevant chunk IDs")
    metadata = JSONField(null=True, help_text="additional context/tags")
    create_time = BigIntegerField(null=False, help_text="creation timestamp")

    class Meta:
        db_table = "evaluation_cases"


class EvaluationRun(DataBaseModel):
    id = CharField(max_length=32, primary_key=True)
    dataset_id = ForeignKeyField(
        EvaluationDataset,
        column_name="dataset_id",
        null=False,
        index=True,
        help_text="FK to evaluation_datasets",
        on_delete="CASCADE",
    )
    dialog_id = ForeignKeyField(
        Dialog,
        column_name="dialog_id",
        null=False,
        index=True,
        help_text="dialog configuration being evaluated",
        on_delete="CASCADE",
    )
    name = CharField(max_length=255, null=False, help_text="run name")
    config_snapshot = JSONField(null=False, help_text="dialog config at time of evaluation")
    metrics_summary = JSONField(null=True, help_text="aggregated metrics")
    status = CharField(max_length=32, null=False, default="PENDING", help_text="PENDING/RUNNING/COMPLETED/FAILED")
    created_by = CharField(max_length=32, null=False, index=True, help_text="user who started the run")
    create_time = BigIntegerField(null=False, index=True, help_text="creation timestamp")
    complete_time = BigIntegerField(null=True, help_text="completion timestamp")

    class Meta:
        db_table = "evaluation_runs"


class EvaluationResult(DataBaseModel):
    id = CharField(max_length=32, primary_key=True)
    run_id = ForeignKeyField(
        EvaluationRun,
        column_name="run_id",
        null=False,
        index=True,
        help_text="FK to evaluation_runs",
        on_delete="CASCADE",
    )
    case_id = ForeignKeyField(
        EvaluationCase,
        column_name="case_id",
        null=False,
        index=True,
        help_text="FK to evaluation_cases",
        on_delete="CASCADE",
    )
    generated_answer = TextField(null=False, help_text="generated answer")
    retrieved_chunks = JSONField(null=False, help_text="chunks that were retrieved")
    metrics = JSONField(null=False, help_text="all computed metrics")
    execution_time = FloatField(null=False, help_text="response time in seconds")
    token_usage = JSONField(null=True, help_text="prompt/completion tokens")
    create_time = BigIntegerField(null=False, help_text="creation timestamp")

    class Meta:
        db_table = "evaluation_results"

File: internal/model/evaluation.go
//
//  Copyright 2026 The InfiniFlow Authors. All Rights Reserved.
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
//

package model

// EvaluationDataset evaluation dataset model
type EvaluationDataset struct {
	ID          string  `gorm:"column:id;primaryKey;size:32" json:"id"`
	TenantID    string  `gorm:"column:tenant_id;size:32;not null;index" json:"tenant_id"`
	Name        string  `gorm:"column:name;size:255;not null;index" json:"name"`
	Description *string `gorm:"column:description;type:longtext" json:"description,omitempty"`
	KbIDs       JSONMap `gorm:"column:kb_ids;type:json;not null" json:"kb_ids"`
	CreatedBy   string  `gorm:"column:created_by;size:32;not null;index" json:"created_by"`
	Status      int64   `gorm:"column:status;default:1;index" json:"status"`
	BaseModel
}

// TableName specify table name
func (EvaluationDataset) TableName() string {
	return "evaluation_datasets"
}

// EvaluationCase evaluation case model
type EvaluationCase struct {
	ID               string   `gorm:"column:id;primaryKey;size:32" json:"id"`
	DatasetID        string   `gorm:"column:dataset_id;size:32;not null;index" json:"dataset_id"`
	Question         string   `gorm:"column:question;type:longtext;not null" json:"question"`
	ReferenceAnswer  *string  `gorm:"column:reference_answer;type:longtext" json:"reference_answer,omitempty"`
	RelevantDocIDs   *JSONMap `gorm:"column:relevant_doc_ids;type:json" json:"relevant_doc_ids,omitempty"`
	RelevantChunkIDs *JSONMap `gorm:"column:relevant_chunk_ids;type:json" json:"relevant_chunk_ids,omitempty"`
	Metadata         *JSONMap `gorm:"column:metadata;type:json" json:"metadata,omitempty"`
	BaseModel
}

// TableName specify table name
func (EvaluationCase) TableName() string {
	return "evaluation_cases"
}

// EvaluationRun evaluation run model
type EvaluationRun struct {
	ID             string   `gorm:"column:id;primaryKey;size:32" json:"id"`
	DatasetID      string   `gorm:"column:dataset_id;size:32;not null;index" json:"dataset_id"`
	DialogID       string   `gorm:"column:dialog_id;size:32;not null;index" json:"dialog_id"`
	Name           string   `gorm:"column:name;size:255;not null" json:"name"`
	ConfigSnapshot JSONMap  `gorm:"column:config_snapshot;type:json;not null" json:"config_snapshot"`
	MetricsSummary *JSONMap `gorm:"column:metrics_summary;type:json" json:"metrics_summary,omitempty"`
	Status         string   `gorm:"column:status;size:32;not null;default:PENDING" json:"status"`
	CreatedBy      string   `gorm:"column:created_by;size:32;not null;index" json:"created_by"`
	BaseModel
}

// TableName specify table name
func (EvaluationRun) TableName() string {
	return "evaluation_runs"
}

// EvaluationResult evaluation result model
type EvaluationResult struct {
	ID              string   `gorm:"column:id;primaryKey;size:32" json:"id"`
	RunID           string   `gorm:"column:run_id;size:32;not null;index" json:"run_id"`
	CaseID          string   `gorm:"column:case_id;size:32;not null;index" json:"case_id"`
	GeneratedAnswer string   `gorm:"column:generated_answer;type:longtext;not null" json:"generated_answer"`
	RetrievedChunks JSONMap  `gorm:"column:retrieved_chunks;type:json;not null" json:"retrieved_chunks"`
	Metrics         JSONMap  `gorm:"column:metrics;type:json;not null" json:"metrics"`
	ExecutionTime   float64  `gorm:"column:execution_time;not null" json:"execution_time"`
	TokenUsage      *JSONMap `gorm:"column:token_usage;type:json" json:"token_usage,omitempty"`
	BaseModel
}

// TableName specify table name
func (EvaluationResult) TableName() string {
	return "evaluation_results"
}
